File: obsidian_excel_automation/utils/date_utils.py
"""
Date utility functions for week/quarter calculations
"""

from datetime import datetime, timedelta
from typing import Tuple


def get_week_number(date: datetime = None, start_day: int = 1) -> int:
    """
    Get ISO week number for a date

    Args:
        date: Date to calculate (defaults to today)
        start_day: 0=Sunday, 1=Monday (default)

    Returns:
        Week number (1-53)
    """
    if date is None:
        date = datetime.now()

    return date.isocalendar()[1]


def get_quarter(date: datetime = None) -> int:
    """
    Get quarter number (1-4) for a date

    Args:
        date: Date to calculate (defaults to today)

    Returns:
        Quarter number (1-4)
    """
    if date is None:
        date = datetime.now()

    return (date.month - 1) // 3 + 1


def get_week_range(week_number: int, year: int = None) -> Tuple[datetime, datetime]:
    """
    Get start and end dates for a week number

    Args:
        week_number: ISO week number
        year: Year (defaults to current year)

    Returns:
        (start_date, end_date) tuple
    """
    if year is None:
        year = datetime.now().year

    # Find first day of the year
    jan1 = datetime(year, 1, 1)

    # Find first Monday (ISO week start)
    days_since_monday = jan1.weekday()
    first_monday = jan1 - timedelta(days=days_since_monday)

    # Calculate week start
    week_start = first_monday + timedelta(weeks=week_number - 1)
    week_end = week_start + timedelta(days=6)

    return week_start, week_end


def get_quarter_range(quarter: int, year: int = None) -> Tuple[datetime, datetime]:
    """
    Get start and end dates for a quarter

    Args:
        quarter: Quarter number (1-4)
        year: Year (defaults to current year)

    Returns:
        (start_date, end_date) tuple
    """
    if year is None:
        year = datetime.now().year

    if quarter not in [1, 2, 3, 4]:
        raise ValueError("Quarter must be 1-4")

    start_month = (quarter - 1) * 3 + 1
    end_month = start_month + 2

    start_date = datetime(year, start_month, 1)

    # Last day of end_month
    if end_month == 12:
        end_date = datetime(year, 12, 31)
    else:
        end_date = datetime(year, end_month + 1, 1) - timedelta(days=1)

    return start_date, end_date


def format_date(date: datetime, format_string: str = "%Y-%m-%d") -> str:
    """
    Format date to string

    Args:
        date: Date to format
        format_string: Python date format string

    Returns:
        Formatted date string
    """
    return date.strftime(format_string)


def parse_date(date_string: str, format_string: str = "%Y-%m-%d") -> datetime:
    """
    Parse date from string

    Args:
        date_string: Date string
        format_string: Python date format string

    Returns:
        Datetime object
    """
    return datetime.strptime(date_string, format_string)


def get_current_week_info() -> dict:
    """Get current week information"""
    now = datetime.now()
    week_num = get_week_number(now)
    start, end = get_week_range(week_num)

    return {
        "week_number": week_num,
        "year": now.year,
        "start_date": start,
        "end_date": end,
        "formatted_week": f"W{week_num}",
        "formatted_date": format_date(now),
    }


def get_current_quarter_info() -> dict:
    """Get current quarter information"""
    now = datetime.now()
    quarter = get_quarter(now)
    start, end = get_quarter_range(quarter)

    return {
        "quarter": quarter,
        "year": now.year,
        "start_date": start,
        "end_date": end,
        "formatted_quarter": f"Q{quarter}",
        "formatted_date": format_date(now),
    }

File: src/utils/dateUtils.ts
/**
 * Date utility functions
 * Equivalent to Python utils/date_utils.py
 */

import { WeekInfo, QuarterInfo } from '../types/models';

/**
 * Get ISO week number for a date
 */
export function getWeekNumber(date: Date = new Date()): number {
  const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
  const dayNum = d.getUTCDay() || 7;
  d.setUTCDate(d.getUTCDate() + 4 - dayNum);
  const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  return Math.ceil((((d.getTime() - yearStart.getTime()) / 86400000) + 1) / 7);
}

/**
 * Get quarter number (1-4) for a date
 */
export function getQuarter(date: Date = new Date()): number {
  return Math.floor(date.getMonth() / 3) + 1;
}

/**
 * Format date according to format string
 * Supports: YYYY, MM, DD, MMM, HH, mm, ss
 */
export function formatDate(date: Date, format: string = 'YYYY-MM-DD'): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  const hours = String(date.getHours()).padStart(2, '0');
  const minutes = String(date.getMinutes()).padStart(2, '0');
  const seconds = String(date.getSeconds()).padStart(2, '0');

  const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  const monthName = monthNames[date.getMonth()];

  return format
    .replace('YYYY', String(year))
    .replace('MMM', monthName)
    .replace('MM', month)
    .replace('DD', day)
    .replace('HH', hours)
    .replace('mm', minutes)
    .replace('ss', seconds);
}

/**
 * Get the date range (start and end) for a given week number
 */
export function getWeekRange(weekNumber: number, year: number): { start: Date; end: Date } {
  // ISO 8601: Week 1 contains Jan 4
  const jan4 = new Date(Date.UTC(year, 0, 4));
  const dayOfWeek = jan4.getUTCDay() || 7; // Mon=1..Sun=7
  const week1Monday = new Date(jan4);
  week1Monday.setUTCDate(jan4.getUTCDate() - (dayOfWeek - 1));

  const start = new Date(week1Monday);
  start.setUTCDate(week1Monday.getUTCDate() + (weekNumber - 1) * 7);
  const end = new Date(start);
  end.setUTCDate(start.getUTCDate() + 6);
  return { start, end };
}

/**
 * Get current week information
 */
export function getCurrentWeekInfo(): WeekInfo {
  const now = new Date();
  const weekNum = getWeekNumber(now);

  return {
    weekNumber: weekNum,
    year: now.getFullYear(),
    formattedWeek: `W${weekNum}`,
    formattedDate: formatDate(now),
  };
}

/**
 * Get current quarter information
 */
export function getCurrentQuarterInfo(): QuarterInfo {
  const now = new Date();
  const quarter = getQuarter(now);

  return {
    quarter,
    year: now.getFullYear(),
    formattedQuarter: `Q${quarter}`,
    formattedDate: formatDate(now),
  };
}

/**
 * Parse date string in various formats
 */
export function parseDate(dateStr: string): Date | null {
  if (!dateStr) return null;

  // Try ISO format: YYYY-MM-DD
  const isoMatch = dateStr.match(/(\d{4})-(\d{2})-(\d{2})/);
  if (isoMatch) {
    return new Date(parseInt(isoMatch[1]), parseInt(isoMatch[2]) - 1, parseInt(isoMatch[3]));
  }

  // Try MM/DD format
  const slashMatch = dateStr.match(/(\d{1,2})\/(\d{1,2})/);
  if (slashMatch) {
    const year = new Date().getFullYear();
    return new Date(year, parseInt(slashMatch[1]) - 1, parseInt(slashMatch[2]));
  }

  return null;
}

/**
 * Check if a date is overdue (before today)
 */
export function isOverdue(date: Date | null): boolean {
  if (!date) return false;
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return date < today;
}

/**
 * Check if a date is within the current week
 */
export function isThisWeek(date: Date | null): boolean {
  if (!date) return false;

  const now = new Date();
  const startOfWeek = new Date(now);
  const dayOfWeek = now.getDay() || 7; // Sunday = 7
  startOfWeek.setDate(now.getDate() - dayOfWeek + 1); // Monday
  startOfWeek.setHours(0, 0, 0, 0);

  const endOfWeek = new Date(startOfWeek);
  endOfWeek.setDate(startOfWeek.getDate() + 6);
  endOfWeek.setHours(23, 59, 59, 999);

  return date >= startOfWeek && date <= endOfWeek;
}
